fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMerger.java

```diff
@@ -0,0 +1,132 @@
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.record.logging.KeyValueLogMessage;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+import com.apple.foundationdb.record.metadata.Index;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.ParametersAreNonnullByDefault;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+/**
+ * Implement an index merge.
+ * This can be either used for an explicit index merge call or during online indexing. In the second case,
+ * keeping throttle values between calls has an overall impact on performance.
+ */
+@ParametersAreNonnullByDefault
+public class IndexingMerger {
+    private static final Logger LOGGER = LoggerFactory.getLogger(IndexingMerger.class);
+    private final Index index;
+    private long mergesLimit;
+    private int mergeSuccesses = 0;
+    private final IndexingCommon common;
+
+    // These error codes represent a list of errors that can occur if there is too much work to be done
+    // in a single transaction.
+    private static final Set<Integer> lessenWorkCodes = new HashSet<>(Arrays.asList(
+            FDBError.TIMED_OUT.code(),
+            FDBError.TRANSACTION_TOO_OLD.code(),
+            FDBError.NOT_COMMITTED.code(),
+            FDBError.TRANSACTION_TIMED_OUT.code(),
+            FDBError.COMMIT_READ_INCOMPLETE.code(),
+            FDBError.TRANSACTION_TOO_LARGE.code()));
+
+    public IndexingMerger(final Index index,  IndexingCommon common, long initialMergesCountLimit) {
+        this.index = index;
+        this.common = common;
+        this.mergesLimit = initialMergesCountLimit;
+    }
+
+    private CompletableFuture<FDBRecordStore> openRecordStore(@Nonnull FDBRecordContext context) {
+        return common.getRecordStoreBuilder().copyBuilder().setContext(context).openAsync();
+    }
+
+    @SuppressWarnings("squid:S3776") // cognitive complexity is high, candidate for refactoring
+    CompletableFuture<Void> mergeIndex() {
+        final AtomicInteger failureCountLimit = new AtomicInteger(1000);
+        AtomicReference<IndexDeferredMaintenancePolicy> policyRef = new AtomicReference<>();
+        return AsyncUtil.whileTrue(() ->
+                common.getRunner().runAsync(context -> openRecordStore(context)
+                                .thenCompose(store -> {
+                                    final IndexDeferredMaintenancePolicy policy = store.getIndexDeferredMaintenancePolicy();
+                                    policyRef.set(policy);
+                                    policy.setMergesLimit(mergesLimit);
+                                    return store.getIndexMaintainer(index).mergeIndex();
```

Comment:
`findMerges` is called by Lucene's `IndexWriter`. I am not too happy about passing arguments and receiving results over the record store object, but it seems that a cleaner way would require Lucene changes. 
If/when we'll create a general `IndexOperations` it will be cleaner, but would probably still be derived - this way or another - from the store. I think.